src/main/java/flink/graphs/example/MusicProfiles.java

```diff
@@ -0,0 +1,181 @@
+	 * on the similarity graph.
+	 */
+	public static void main (String [] args) throws Exception {
+    	
+    	ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+    	final long numberOfLabels = 3;
+    	final int numIterations = 10;
+
+    	/** 
+    	 *  Read the user-song-play triplets
+    	 *  The format is <userID>\t<songID>\t<playcount>
+    	 */
+    	DataSet<Tuple3<String, String, Integer>> triplets = MusicProfilesData.getUserSongTriplets(env);
+
+    	/**
+    	 *  Read the mismatches dataset and extract the songIDs
+    	 *  The format is "ERROR: <songID trackID> song_title"
+    	 */
+    	DataSet<Tuple1<String>> mismatches = MusicProfilesData.getMismatches(env).map(new ExtractMismatchSongIds());
+
+    	/**
+    	 *  Filter out the mismatches from the triplets dataset
+    	 */
+    	DataSet<Tuple3<String, String, Integer>> validTriplets = triplets.coGroup(mismatches)
+    			.where(1).equalTo(0).with(new FilterOutMismatches());
+
+    	/**
+    	 *  Create a user -> song weighted bipartite graph
+    	 *  where the edge weights correspond to play counts
+    	 */
+    	DataSet<Edge<String, Integer>> userSongEdges = validTriplets.map(new Tuple3ToEdgeMap<String, Integer>());
+    	Graph<String, NullValue, Integer> userSongGraph = Graph.create(userSongEdges, env);
+
+    	/**
+    	 *  Get the top track (most listened) for each user
+    	 */
+    	DataSet<Tuple2<String, String>> usersWithTopTrack = userSongGraph.reduceOnEdges(new GetTopSongPerUser(), 
+    			EdgeDirection.OUT).filter(new FilterSongNodes());
+    	usersWithTopTrack.print();
+
+    	/**
+    	 * Create a user-user similarity graph, based on common songs, 
+    	 * i.e. two users that listen to the same song are connected.
+    	 * For each song, we create an edge between each pair of its in-neighbors.
+    	 */
+    	DataSet<Edge<String, NullValue>> similarUsers = userSongGraph.getEdges().groupBy(1)
+    			.reduceGroup(new CreateSimilarUserEdges()).distinct();
+    	Graph<String, NullValue, NullValue> similarUsersGraph = Graph.create(similarUsers, env).getUndirected();
+
+    	/**
+    	 * Detect user communities using the label propagation library method
+    	 */
+    	DataSet<Vertex<String, Long>> verticesWithCommunity = similarUsersGraph.mapVertices(
+    			new InitVertexLabels(numberOfLabels))
+    			.run(new LabelPropagation<String>(numIterations)).getVertices();
+    	verticesWithCommunity.print();
+
+    	env.execute();
+    }
+
+    @SuppressWarnings("serial")
+	public static final class ExtractMismatchSongIds implements MapFunction<String, Tuple1<String>> {
+		public Tuple1<String> map(String value) {
+			String[] tokens = value.split("\\s+"); 
+			String songId = tokens[1].substring(1);
+			return new Tuple1<String>(songId);
+		}
+    }
+
+    @SuppressWarnings("serial")
+	public static final class FilterOutMismatches implements CoGroupFunction<Tuple3<String, String, Integer>, 
```

Comment:
This is actually an *anti-join* like in a *where not exists* subquery. We should add this to the APIs, it seems a common pattern.